blog/2024-11-19-microsoft-azure-authentication.md

```diff
@@ -0,0 +1,348 @@
+        }
+        try:
+            result = sso_app.acquire_token_by_auth_code_flow(
+                self._flow, auth_response, scopes=[]
+            )
+        except Exception as e:
+            return rx.toast(f"error something went wrong")
+        # this can be used for accessing graph
+        self._access_token = result.get("access_token")
+        self._token = result.get("id_token_claims")
+        return rx.redirect(login_redirect)
+
+
+
+@rx.page(route="/callback", on_load=State.callback)
+def callback() -> rx.Component:
+    return rx.container()
+```
+
+
+## Logging into the App
+
+Now that we have all our tokens set correctly, when we redirect back to the home page, again the `require_auth` event handler is called. This time, the `_token` variable is not empty and so the page renders.
+
+The page contains an [`rx.cond`]({docs.components.conditional_rendering.path}) component that checks if the user is authenticated. It runs the `check_auth` [computed var]({docs.vars.computed_vars.path}), which returns `True` if the `_token` variable is not empty and `False` otherwise.
+
+If the user is authenticated, the `auth_view` component is rendered. If the user is not authenticated, the `unauth_view` component is rendered. The `auth_view` component displays a welcome message with the user's name from the `token` [computed var]({docs.vars.computed_vars.path}), which returns the `_token` [backend variable]({docs.vars.base_vars.path}#backend-only-vars).
+
+
+```python
+class State(rx.State):
+    _access_token: str = ""
+    _flow: dict
+    _token: Dict[str, str] = \{}
+
+    def require_auth(self):
+        if not self._token:
+            return self.redirect_sso()
+
+    @rx.var(cache=True)
+    def check_auth(self) -> bool:
+        return True if self._token else False
+
+    @rx.var(cache=True)
+    def token(self) -> Dict[str, str]:
+        return self._token
+
+
+
+@rx.page(route="/", on_load=State.require_auth)
+def home() -> rx.Component:
+    return rx.container(rx.cond(State.check_auth, auth_view(), unauth_view()))
+
+
+def auth_view() -> rx.Component:
+    return rx.text(f"Hello \{State.token['name']}")
+
+
+def unauth_view() -> rx.Component:
+    return rx.text("Unauthorized, redirected...")
+```
+
+
+## Logging Out
+
+Lastly we have the logout functionality. When the user goes to the `/logout` page the `logout` event handler is called. This event handler clears the `_token` variable and redirects the user to the Microsoft logout page. This event handler could also easliy be put on the `on_click` event of a button.
+
+```python
+class State(rx.State):
+    _access_token: str = "" 
+    _flow: dict 
+    _token: Dict[str, str] = \{}
+
+    def logout(self):
+        self._token = \{}
+        return rx.redirect(authority + "/oauth2/v2.0/logout")
+
+
+@rx.page(route="/logout", on_load=State.logout)
+def logout() -> rx.Component:
+    return rx.container("Logged out")
+```
+
+## Putting it all together
```

Comment:
maybe link to the examples git for this as well?

Reply:
added